gold-phone-input.html

```diff
@@ -181,51 +194,122 @@
-      if (oldValue == undefined || value === oldValue)
+      if (oldValue == undefined || value == undefined) {
+        this.value = '';
         return;
+      }
 
-      //Ensure value is a string
       value = value ? value.toString() : '';
 
       // Keep track of how many dashes the original value has. After
       // reformatting the value, we might gain or lose some of them, which
       // means we have to correctly move the caret to account for the difference.
       var start = this.$.input.selectionStart;
+
       var initialDashesBeforeCaret = value.substr(0, start).split('-').length - 1;
+      var initialSpacesBeforeCaret = value.substr(0, start).split(' ').length - 1;
+      var initialLeftParenBeforeCaret = value.substr(0, start).split('(').length - 1;
+      var initialRightParenBeforeCaret = value.substr(0, start).split(')').length - 1;
+
+      var formattedValue = '';
 
       // Remove any already-applied formatting.
       value = value.replace(/-/g, '');
-      var shouldFormat = value.length <= this.phoneNumberPattern.replace(/-/g, '').length;
-      var formattedValue = '';
+      value = value.replace(/\(/g, '');
+      value = value.replace(/\)/g, '');
+      value = value.replace(/\s/g, '');
+
+      oldValue = oldValue.replace(/-/g, '');
+      oldValue = oldValue.replace(/\(/g, '');
+      oldValue = oldValue.replace(/\)/g, '');
+      oldValue = oldValue.replace(/\s/g, '');
+
+      if(value === oldValue) {
+        return;
+      }
+
+      if(value.length > 10) {
+        this.updateValueAndPreserveCaret(value.trim());
+        this._handleAutoValidate();
+        return;
+      }
 
       // Fill in the dashes according to the specified pattern.
+      var leftParenIndex = this.phoneNumberPattern.indexOf('(');
+      var leftParenAdded = false;
+
+      var rightParenIndex = this.phoneNumberPattern.indexOf(')') - 1;
+      var rightParenAdded = false;
+
+      var currentSpaceIndex = 0;
       var currentDashIndex = 0;
-      var totalDashesAdded = 0;
+
+      var totalCharactersAdded = 0;
+
       for (var i = 0; i < value.length; i++) {
+
         currentDashIndex = this.phoneNumberPattern.indexOf('-', currentDashIndex);
 
-        // Since we remove any formatting first, we need to account added dashes
+        currentSpaceIndex = this.phoneNumberPattern.indexOf(' ', currentSpaceIndex);
+
+        // Since we remove any formatting first, we need to account added dashes, etc.
         // when counting the position of new dashes in the pattern.
-        if (shouldFormat && i == (currentDashIndex - totalDashesAdded)) {
+
+        if (i == leftParenIndex) {
+          formattedValue += '(';
+          leftParenAdded = true;
+          totalCharactersAdded++;
+        }
+
+        if (i == rightParenIndex) {
+          formattedValue += ')';
+          rightParenAdded = true;
+          totalCharactersAdded++;
+        }
+
+        if (i == (currentSpaceIndex - totalCharactersAdded)) {
```

Comment:
This needs a lot of comments. At the moment, it's very hard to understand what this code is doing